cpp/src/search/contains_table.cu

```diff
@@ -173,124 +200,106 @@ rmm::device_uvector<bool> contains(table_view const& haystack,
+        cuco::empty_key{lhs_index_type{-1}},
+        d_equal,
+        probing_scheme,
+        detail::hash_table_allocator_type{default_allocator<lhs_index_type>{}, stream},
+        stream.value()};
+
+      if (haystack_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+        auto const bitmask_buffer_and_ptr = build_row_bitmask(haystack, stream);
+        auto const row_bitmask_ptr        = bitmask_buffer_and_ptr.second;
+
+        // If the haystack table has nulls but they are compared unequal, don't insert them.
+        // Otherwise, it was known to cause performance issue:
+        // - https://github.com/rapidsai/cudf/pull/6943
+        // - https://github.com/rapidsai/cudf/pull/8277
+        set.insert_if_async(haystack_iter,
+                            haystack_iter + haystack.num_rows(),
+                            thrust::counting_iterator<size_type>(0),  // stencil
+                            row_is_valid{row_bitmask_ptr},
+                            stream.value());
+      } else {
+        set.insert_async(haystack_iter, haystack_iter + haystack.num_rows(), stream.value());
+      }
+
+      if (needles_has_nulls && compare_nulls == null_equality::UNEQUAL) {
+        auto const bitmask_buffer_and_ptr = build_row_bitmask(needles, stream);
+        auto const row_bitmask_ptr        = bitmask_buffer_and_ptr.second;
+        set.contains_if_async(needles_iter,
+                              needles_iter + needles.num_rows(),
+                              thrust::counting_iterator<size_type>(0),  // stencil
+                              row_is_valid{row_bitmask_ptr},
+                              contained.begin(),
+                              stream.value());
       } else {
-        auto const d_eqcomp =
-          comparator.equal_to<false>(nullate::DYNAMIC{has_any_nulls}, compare_nulls, value_comp);
-        map.contains(needles_it,
-                     needles_it + needles.num_rows(),
-                     contained.begin(),
-                     d_hasher,
-                     d_eqcomp,
-                     stream.value());
+        set.contains_async(
+          needles_iter, needles_iter + needles.num_rows(), contained.begin(), stream.value());
       }
     };
 
-    dispatch_nan_comparator(compare_nans, check_contains);
+  if (cudf::detail::has_nested_columns(haystack)) {
+    dispatch_nan_comparator<true, true>(compare_nulls,
+                                        compare_nans,
+                                        haystack_has_nulls,
+                                        has_any_nulls,
+                                        self_equal,
+                                        two_table_equal,
+                                        d_hasher,
+                                        helper_func);
+  } else {
+    if (cudf::detail::has_nested_columns(needles)) {
+      dispatch_nan_comparator<false, true>(compare_nulls,
```

Comment:
This is the case I'm interested in investigating in the other thread. Is the `<false, true>` dispatch providing special functionality or performance advantages that can't be achieved with 2 dispatches (`<true, true>` or `<false, false>`)?

Reply:
Yeah, I see your point. 

Ideally, users should always check nested against nested. Having the `<false, true>` dispatch allows users to check nested needles against a flat haystack (though I'm not sure either if it's a legit use case), and the haystack set insertion would be faster than the `<true, true>` dispatch.

For my education, in this particular case, I assume `<true, true>` would work fine but less efficiently since we do redundant nested-related checks for a flat column. But would `<false, false>` produce proper results?

Reply:
>  But would <false, false> produce proper results?

Checked with @divyegala, this will cause runtime errors.

Reply:
> Having the `<false, true>` dispatch allows users to check nested needles against a flat haystack (though I'm not sure either if it's a legit use case)

Maybe I'm misunderstanding the problem -- but it seems like a nested needle could never be found in a haystack without nesting. Shouldn't that fail because the types of the needle and haystack don't match?

Reply:
> Shouldn't that fail because the types of the needle and haystack don't match?

I'm not sure. The current `contains_table` allows this combination and returns 0 in this case. I'm trying to align with the existing behavior. Maybe @ttnghia knows more about what is the expected behavior here.

Reply:
Sorry, I shouldn't approve this. I almost forgot about the behavior of the comparator regarding the `has_nulls`.

Here is the thing: when the column has nulls, the validity (true/false) of the column will be hashed into the output hash:
```
if (_check_nulls) {
          auto validity_it = detail::make_validity_iterator<true>(curr_col);
          hash             = detail::accumulate(..., hash,...);
```

When using hash table (map/set), we must ensure that the hash values for rows in both tables are consistent. That means, if one side table has nulls and hashed with nulls, the other side must also be hashed with the same `has_nulls`. Otherwise, one table is hashed differently than the other and you will not find the correct "contains" output.

In summary, the same `hash_nulls` value should be used for both inserting `lhs` and finding contains of `rhs`.

Reply:
I think that's a different issue. Null handling is properly propagated to both self and two-table comparators.

The focus was whether it's legit to have mismatched column types between haystack and needles. After checking the public `contains` API, I think the answer is clearer now: I've updated the implementation so now it will throw if column types mismatch. The implementation is therefore simplified as @bdice suggested. Nice catch, thanks! :pray: